File: code/com/leetcode/solutions/NO_0012_整数转罗马数字/Solution.cs
using System;

public class Solution
{
	
    /*
    * num:指定数字
    * 思路：分别求出各个位上的数字，然后进行判断。
    * 其他：将罗马数字分别定义为四个数组，然后得到位数之后对应去取。
   */

        public string IntToRoman(int num)
        {
            //千位
            int s = num / 1000;
            int t = (num % 1000) / 100;
            int v = ((num % 1000) % 100) / 10;
            int k = ((num % 1000) % 100) % 10;
            int[] arr = { s, t, v, k };
            StringBuilder res = new StringBuilder();
            for (int i = 0; i < arr.Length; i++)
            {
                int item = arr[i];
                switch (i)
                {
                    case 0:
                        // 千位
                        if (item > 0)
                        {
                            for (int j = 0; j < item; j++)
                            { res.Append("M"); }
                        }
                        break;
                    case 1:
                        //百位
                        if (item > 0 && item < 4)
                        {
                            for (int j = 0; j < item; j++)
                            { res.Append("C"); }
                        }
                        else if (item == 4)
                        {
                            res.Append("CD");
                        }
                        else if (item == 5)
                        {
                            res.Append("D");
                        }
                        else if (item > 5 && item < 9)
                        {
                            res.Append("D");
                            for (int j = 0; j < item - 5; j++)
                            { res.Append("C"); }
                        }
                        else if (item == 9)
                        {
                            res.Append("CM");
                        }
                        break;
                    case 2:
                        // 十位
                        if (item > 0 && item < 4)
                        {
                            for (int j = 0; j < item; j++)
                            { res.Append("X"); }
                        }
                        else if (item == 4)
                        {
                            res.Append("XL");
                        }
                        else if (item == 5)
                        {
                            res.Append("L");
                        }
                        else if (item > 5 && item < 9)
                        {
                            res.Append("L");
                            for (int j = 0; j < item - 5; j++)
                            { res.Append("X"); }
                        }
                        else if (item == 9)
                        {
                            res.Append("XC");
                        }
                        break;

                    case 3:
                        if (item > 0 && item < 4)
                        {
                            for (int j = 0; j < item; j++)
                            { res.Append("I"); }
                        }
                        else if (item == 4)
                        {
                            res.Append("IV");
                        }
                        else if (item == 5)
                        {
                            res.Append("V");
                        }
                        else if (item > 5 && item < 9)
                        {
                            res.Append("V");
                            for (int j = 0; j < item - 5; j++)
                            { res.Append("I"); }
                        }
                        else if (item == 9)
                        {
                            res.Append("IX");
                        }
                        break;
                }
            }
            return res.ToString();
        }
  
}

File: code/com/leetcode/solutions/NO_0012_整数转罗马数字/Solution.java
package com.leetcode.solutions.NO_0012_整数转罗马数字;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Created by gang.xu01@hand-china.com on 2019/1/13
 */
public class Solution {


    private final LinkedHashMap<Integer, String> format = new LinkedHashMap<Integer, String>(){{
        put(1000, "M");
        put(900, "CM");
        put(500, "D");
        put(400, "CD");
        put(100, "C");
        put(90, "XC");
        put(50, "L");
        put(40, "XL");
        put(10, "X");
        put(9, "IX");
        put(5, "V");
        put(4, "IV");
        put(1, "I");
    }};

    /**
     * 解法：一层一层遍历得到对应的罗马数字即可
     *
     * @param num
     * @return
     */
    public String intToRoman(int num) {
        if (num < 1 || num > 3999) throw new RuntimeException("Not in range");

        StringBuilder sb = new StringBuilder();
        Map<String, Integer> map = new HashMap<String, Integer>(){{
            put("value", num);
        }};

        format.forEach( (k, v) -> {
            int n = map.get("value");
            int t = n / k;
            if (t > 0) {
                map.put("value", n % k);
                for (int i = 0; i < t; i++) sb.append(v);
            }
        });

        return sb.toString();
    }

    public static void main(String[] args) {
        System.out.println(new Solution().intToRoman(1994));
//        System.out.println( 994 % 900);
    }

}
